### DISCUSSION TITLE
Merge queue force-pushes stacked PRs upon merge

### DISCUSSION THREAD

**Discussion by MarijnS95** [community/community#38167] [Pull Requests]
This is a bug report.

### Context

We use the Merge Queue beta, together with squash merges.

When we have more than one PR in the merge queue, GitHub neatly squash-merges them all on top of main, and pushes the result to `gh-readonly-queue/main/pr-xxxx-yyyy` branches, where the CI runs on all of them in parallel.

### Expected outcome

As soon as CI finishes on the first PR/branch in the queue, `main` is fast-forwarded to it. Whenever CI finishes on the other squash-merged PRs on these branches, `main` is fast-forwarded to those. This used to work fine and meant that multiple PRs could be merged in about the same time as one PR, assuming enough CI resources are available to check multiple PRs/commits in parallel.

### The problem

Since about 2-3 weeks the merge queue _force pushes_ all the `gh-readonly-queue/main/pr-xxxx-yyyy` branches as soon as _the first_ commit in the queue succeeds CI and is "merged" into `main` (more specifically: `main` is fast-forwarded to this commit). This means all CI progress made on the other commits/PRs _is now lost_, despite these force-pushed commits carrying the _exact same contents_ and parent commit hash, only the timestamp (hence resulting commit hash) changed. This _wastes_ CI time, and time it takes to merge a PR, scaling logarithmically based on queue length.

EDIT: Made worse: the CI might still be processing the original commits, unnecessarily keeping it occupied when the result is discarded _anyway_.

### Expected solution

The merge queue stops force-pushing these branches after initially stacking multiple squash-merged PRs, _unless_ it had to kick an earlier PR out of the queue because of failed status checks or user request.

### Additional symptoms

One strange feat we noticed is with our Slack GitHub integration: for some of these `gh-readonly-queue/main/pr-xxxx-yyyy` branches it only ever shows `` force-pushed the `gh-readonly-queue/main/pr-xxxx-yyyy` branch from `uuuu` to `vvvv` ``. Here commit `uuuu` is always a _merge_ commit with _two parents_ (one is the `HEAD` of PR `xxxx`, the other is an earlier squash-merged commit in the queue that `main` was just fast-forwarded to, whose hash is actually encoded in the name of the branch as `yyyy` :slightly_smiling_face:), despite us _not having merge commits enabled at all_.
However, our Slack GitHub integration never printed the original `` 1 new commit (force-?)pushed to `gh-readonly-queue/main/pr-xxxx-yyyy` by ZZZ ``, and I have never been able to fetch the branch while it is in this state, as if the branch was previously hidden while GitHub originally pushed this merge commit to it?

**Comment by mrmckeb** [comment]
Has anyone found a temporary workaround to this issue? Like @MarijnS95, we're finding that this is consuming a lot of extra resource - on top of it not working as expected, and as it previously did.

**Reply by haugene** [reply]
We're using [skip-duplicate-actions](https://github.com/marketplace/actions/skip-duplicate-actions) to skip most of our CI jobs on main branch when they have been run in the merge queue. It compares the treeHash so it will correctly identify that the contents of the commit has not changed.

This is a Python project and our jobs consist of static code checks and running tests + two docker image builds. We skip the code checks and tests but still perform the docker image builds because we need to have an image with the same tag as the git commit sha for the deployment pipeline. We are using the docker/build-push-action with build caching so rebuilding the image is also fast (< 1 min) as Docker hits all the cache layers.

**Comment by parched** [comment]
> Since about 2-3 weeks the merge queue force pushes all the gh-readonly-queue/main/pr-xxxx-yyyy branches as soon as the first commit in the queue succeeds CI and is "merged" into main (more specifically: main is fast-forwarded to this commit)

This force push isn't happening to us. However, a new `merge_group` workflow run is being started with the exact same `ref` and `sha`. Which, again, wastes runner and developer time.
